### PR TITLE
ATRC survey block validation

### PR DESCRIPTION
**Pull request by nspool**
For ATRC surveys validate that m1 has more than one record for block 1 and 2 and that m2 has more that one for block 1.

### CODE REVIEW

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java [hunk 1]
```diff
@@ -30,34 +33,88 @@ public ATRCSurveyGroupComplete() {
...
+
+        List<Validated<StagedRowError,String>> result = new ArrayList<Validated<StagedRowError,String>>();
+
+        surveys.keySet().stream().forEach(key -> {
+
+            List<StagedSurveyTransect> surveyRecords = surveys.get(key);
+            Map<String, List<StagedSurveyTransect>> surveyByMethod = surveyRecords.stream()
+                    .collect(Collectors.groupingBy(StagedSurveyTransect::getMethod));
+
+            surveyByMethod.keySet().stream().forEach(method -> {
```
Comment (jonescc) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java:65]: Perhaps just check methods 1 and 2 specifically - there are other combinations - and there are no rules specified for those methods:

```
select distinct sm.method_id, block_num
  from nrmn.survey_method sm
  join nrmn.survey s on (s.survey_id = sm.survey_id and s.program_id = 2)
```

method_id | block_num
-- | --
3 | 0
7 | 2
10 | 1
1 | 1
11 | NULL
4 | NULL
2 | 2
0 | NULL
6 | NULL
2 | 1
3 | NULL
5 | NULL
10 | 2
1 | 2


@bpasquer - can you confirm.

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java [hunk 2]
```diff
@@ -30,34 +33,88 @@ public ATRCSurveyGroupComplete() {
...
+
+        return result.stream().reduce(Validated.valid(""),
+                (acc, validator) -> acc.combine(Monoids.stringConcat, validator));
     }
 
     private boolean surveyGroupComplete(List<String> surveyNums) {
         return surveyNums.containsAll(Arrays.asList("1", "2", "3", "4"));
     }
+
+    private Validated<StagedRowError, String> surveyMethodBlocksComplete(long jobId,
```
Comment (jonescc) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java:89]: Not used/duplicate code?

**Comment by nspool** [comment]
Deployed stack to: http://nrmn-atrc-survey-block.dev.aodn.org.au/

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java [hunk 3]
```diff
@@ -30,31 +33,59 @@ public ATRCSurveyGroupComplete() {
...
+
+            List<StagedSurveyTransect> surveyRecords = surveys.get(key);
+            Map<String, List<StagedSurveyTransect>> surveyByMethod = surveyRecords.stream()
+                    .collect(Collectors.groupingBy(StagedSurveyTransect::getMethod));
+
+            surveyByMethod.keySet().stream().forEach(method -> {
+
+                List<StagedSurveyTransect> rows = surveyByMethod.get(method);
+                String surveyName = rows.get(0).getDepth() + "." + rows.get(0).getSurveyNum();
+                String errorMessage = "Survey " + surveyName + " is missing Method " + method;
```
Comment (jonescc) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java:70]: This is probably not enough information for them to be able to find the problem survey - the spreadsheet can contain more than one site/date.

Comment (nspool) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/validators/global/raw/ATRCSurveyGroupComplete.java]: Will now display the site code/date/survey at the start of the message.

**Pull request merged**